File: docx_parsers/models/paragraph_models.py
from pydantic import BaseModel, Field
from typing import List, Optional, Union
from docx_parsers.models.styles_models import ParagraphStyleProperties, RunStyleProperties

class Numbering(BaseModel):
    """
    Represents the numbering properties of a paragraph.

    Example:
        The following is an example of a numbering element in a paragraph properties element:

        .. code-block:: xml

            <w:numPr>
                <w:ilvl w:val="0"/>
                <w:numId w:val="1"/>
            </w:numPr>
    """
    ilvl: int = Field(..., description="The indent level of the numbering.")
    numId: int = Field(..., description="The ID of the numbering definition.")

class TextContent(BaseModel):
    """
    Represents text content in a run.

    Example:
        The following is an example of a text element in a run:

        .. code-block:: xml

            <w:r>
                <w:t>Example text</w:t>
            </w:r>
    """
    text: str = Field(..., description="The text content.")

class TabContent(BaseModel):
    """
    Represents a tab character in a run.

    Example:
        The following is an example of a tab element in a run:

        .. code-block:: xml

            <w:r>
                <w:tab/>
            </w:r>
    """
    type: str = Field("tab", description="The type of content, default is 'tab'.")

class RunContent(BaseModel):
    """
    Represents the content of a run, which can be either text or a tab.

    Example:
        The following is an example of run contents in a run element:

        .. code-block:: xml

            <w:r>
                <w:t>Example text</w:t>
                <w:tab/>
            </w:r>
    """
    run: Union[TextContent, TabContent] = Field(..., description="The content of the run.")

class Run(BaseModel):
    """
    Represents a run within a paragraph, containing text and formatting properties.

    Example:
        The following is an example of a run element in a paragraph:

        .. code-block:: xml

            <w:r>
                <w:rPr>
                    <w:b/>
                    <w:color w:val="FF0000"/>
                </w:rPr>
                <w:t>Example text</w:t>
            </w:r>
    """
    contents: List[RunContent] = Field(..., description="The list of run contents (text or tabs).")
    properties: Optional[RunStyleProperties] = Field(None, description="The style properties of the run.")

class Paragraph(BaseModel):
    """
    Represents a paragraph in the document, containing text runs and optional numbering.

    Example:
        The following is an example of a paragraph element in a document:

        .. code-block:: xml

            <w:p>
                <w:pPr>
                    <w:pStyle w:val="Heading1"/>
                    <w:numPr>
                        <w:ilvl w:val="0"/>
                        <w:numId w:val="1"/>
                    </w:numPr>
                </w:pPr>
                <w:r>
                    <w:t>Example text</w:t>
                </w:r>
            </w:p>
    """
    properties: ParagraphStyleProperties = Field(..., description="The style properties of the paragraph.")
    runs: List[Run] = Field(..., description="The list of text runs within the paragraph.")
    numbering: Optional[Numbering] = Field(None, description="The numbering properties, if the paragraph is part of a list.")

File: docx_parser_converter_ts/src/docx_parsers/models/paragraphModels.ts
import type { ParagraphStyleProperties, RunStyleProperties } from './stylesModels';

/**
 * Represents the numbering properties of a paragraph.
 * 
 * @example
 * The following is an example of a numbering element in a paragraph properties element:
 * ```xml
 * <w:numPr>
 *     <w:ilvl w:val="0"/>
 *     <w:numId w:val="1"/>
 * </w:numPr>
 * ```
 */
export interface Numbering {
  /** The indent level of the numbering. */
  ilvl: number;
  /** The ID of the numbering definition. */
  numId: number;
}

/**
 * Represents text content in a run.
 * 
 * @example
 * The following is an example of a text element in a run:
 * ```xml
 * <w:r>
 *     <w:t>Example text</w:t>
 * </w:r>
 * ```
 */
export interface TextContent {
  /** The text content. */
  text: string;
  /** Content type discriminator. */
  type: 'text';
}

/**
 * Represents a tab character in a run.
 * 
 * @example
 * The following is an example of a tab element in a run:
 * ```xml
 * <w:r>
 *     <w:tab/>
 * </w:r>
 * ```
 */
export interface TabContent {
  /** The type of content, default is 'tab'. */
  type: 'tab';
}

/**
 * Represents the content of a run, which can be either text or a tab.
 * 
 * @example
 * The following is an example of run contents in a run element:
 * ```xml
 * <w:r>
 *     <w:t>Example text</w:t>
 *     <w:tab/>
 * </w:r>
 * ```
 */
export type RunContent = TextContent | TabContent;

/**
 * Represents a run within a paragraph, containing text and formatting properties.
 * 
 * @example
 * The following is an example of a run element in a paragraph:
 * ```xml
 * <w:r>
 *     <w:rPr>
 *         <w:b/>
 *         <w:color w:val="FF0000"/>
 *     </w:rPr>
 *     <w:t>Example text</w:t>
 * </w:r>
 * ```
 */
export interface Run {
  /** The list of run contents (text or tabs). */
  contents: RunContent[];
  /** The style properties of the run. */
  properties?: RunStyleProperties;
}

/**
 * Represents a paragraph in the document, containing text runs and optional numbering.
 * 
 * @example
 * The following is an example of a paragraph element in a document:
 * ```xml
 * <w:p>
 *     <w:pPr>
 *         <w:pStyle w:val="Heading1"/>
 *         <w:numPr>
 *             <w:ilvl w:val="0"/>
 *             <w:numId w:val="1"/>
 *         </w:numPr>
 *     </w:pPr>
 *     <w:r>
 *         <w:t>Example text</w:t>
 *     </w:r>
 * </w:p>
 * ```
 */
export interface Paragraph {
  /** The style properties of the paragraph. */
  properties: ParagraphStyleProperties;
  /** The list of text runs within the paragraph. */
  runs: Run[];
  /** The numbering properties, if the paragraph is part of a list. */
  numbering?: Numbering;
} 